src/daemon/backends/connman_agent.cpp

```diff
@@ -0,0 +1,336 @@
+                        g_warning("Received ConnMan agent fields with both %s and %s",
+                                  FIELD_NAME_EAP_USERNAME,
+                                  FIELD_NAME_WISPR_USERNAME);
+                        return {};
+                    }
+                    credentials.username = REQUESTED_BUT_NO_PREVIOUS_VALUE_EVER_INCLUDED;
+
+                } else if (name == FIELD_NAME_PASSPHRASE || name == FIELD_NAME_WISPR_PASSWORD) {
+                    if (credentials.password) {
+                        g_warning("Received ConnMan agent fields with both %s and %s",
+                                  FIELD_NAME_PASSPHRASE,
+                                  FIELD_NAME_WISPR_PASSWORD);
+                        return {};
+                    }
+                    credentials.password = value_to_password(name, value);
+                    if (!credentials.password)
+                        return {};
+
+                } else if (name == FIELD_NAME_PREVIOUS_PASSPHRASE) {
+                    previous_password = value_to_password(name, value);
+                    if (!previous_password)
+                        return {};
+
+                } else if (name == FIELD_NAME_WPS) {
+                    credentials.password_alternative = value_to_password(name, value);
+
+                    if (!credentials.password_alternative ||
+                        credentials.password_alternative->type != Password::Type::WPS_PIN) {
+                        g_warning("Received ConnMan agent WPS field with wrong type");
+                        return {};
+                    }
+
+                } else {
+                    g_warning("Received unknown ConnMan agent field \"%s\"", name.c_str());
+                    return {};
+                }
+            }
+
+            if (credentials.password_alternative) {
+                if (!credentials.password) {
+                    g_warning("Received ConnMan agent fields with password alternative field and "
+                              "no password field");
+                    return {};
+                }
+
+                if (credentials.password->type == credentials.password_alternative->type) {
+                    g_warning("Received ConnMan agent fields with password and password "
+                              "alternative of same type");
+                    return {};
+                }
+            }
+
+            if (previous_password) {
+                if (!credentials.password) {
+                    g_warning("Received ConnMan agent fields with previous password field and no "
+                              "password field");
+                    return {};
+                }
+
+                if (credentials.password->type == previous_password->type) {
+                    if (credentials.password->value.empty())
+                        credentials.password->value = previous_password->value;
+
+                } else if (credentials.password_alternative &&
```

Comment:
It looks like it does here since the ConnMan agent API is underspecified... and bad. There is a way to have "values" in "fields". They could have used that for passing any previous password. But instead they added another field for previous password.

Perhaps I do not understand what you mean though.

Reply:
Hmm so I guess I was a bit confused ... It is more clear when I re-read the code and I seem to have misunderstood it .

Reply:
Maybe it could be broken out in to functions that handle different cases to make it easier to read? It is hard to say how this can be done in a more readable way

Reply:
Perhaps this warrants some documentation at the start of the function... since the D-Bus API is so... wönderfül. I scratched my head over https://git.kernel.org/pub/scm/network/connman/connman.git/tree/doc/agent-api.txt#n97 ... a lot.

Reply:
Added a short blurb at top of function trying to explain the "Value" vs "PreviousPassphrase" thing and that there are a couple of examples in doc/agent-api.txt . Suggestions for how to improve it further welcome... I think I am to tired to come up with anything better at the moment.

Reply:
Resolving. Please reopen/comment again if you have any ideas for how to improve it.